### PR TITLE
Add user-defined callbacks.

### PR DESCRIPTION
**Pull request by sile**
This PR adds `callbacks` argument to `study.optimize()` method.
The passed callbacks are invoked just after each trial finishes.

### Example

```python
def trial_log_callback(study, trial):
    print("TRIAL_NO={}, PARAMS={}, VALUE={}".format(trial.number, trial.params, trial.value))

study = optuna.create_study()
study.optimize(objective, n_trials=100, callbacks = [trial_log_callback])
```

### Related Issues

- #302 
- #389

### CODE REVIEW

**Comment by sile** [comment]
### Example: TensorBoard

```python
import optuna
from tensorboardX import SummaryWriter


# Callback.
def update_tensorboardx(writer, study, trial):
    writer.add_scalar('data/value', trial.value, trial.number)
    writer.add_scalar('data/best_value', study.best_value, trial.number)


def objective(trial):
    x = trial.suggest_uniform('x', -100, 100)
    y = trial.suggest_categorical('y', [-1, 0, 1])
    return x**2 + y


if __name__ == '__main__':
    study = optuna.create_study()
    writer = SummaryWriter()
    study.optimize(objective, timeout=300,
                   callbacks=[lambda x, y: update_tensorboardx(writer, x, y)])
    writer.close()
```

image

**Comment by sile** [comment]
### Example: Neptune

```python
import optuna
import os
import neptune


# Callback.
def report_neptune(study, trial):
    neptune.send_metric('value', trial.value)
    neptune.send_metric('best_value', study.best_value)


def objective(trial):
    x = trial.suggest_uniform('x', -100, 100)
    y = trial.suggest_categorical('y', [-1, 0, 1])
    return x**2 + y


if __name__ == '__main__':
    # Setup neptune.ml.
    neptune.init('{}/sandbox'.format(os.environ['NEPTUNE_USER']))
    neptune.create_experiment()

    # Let us minimize the objective function above.
    study = optuna.create_study()
    study.optimize(objective, n_trials=10, callbacks=[report_neptune])
    neptune.stop()
```

image

File: optuna/study.py [hunk 1]
```diff
@@ -437,6 +441,20 @@ def func_child_thread(que):
         que.close()
         que.join_thread()
 
+    def _run_trial_and_callbacks(
```
Comment (cthoyt) [inline optuna/study.py]: Really elegant implementation. Nice.

File: tests/test_study.py [hunk 2]
```diff
@@ -559,3 +559,58 @@ def test_storage_property():
...
+        return trial.suggest_int('x', 1, 1)
+
+    # Empty callback list.
+    study.optimize(objective, callbacks=[], n_trials=10, n_jobs=n_jobs)
+
+    # A callback.
+    values = []
+    callbacks = [lambda study, trial: values.append(trial.value)]
+    study.optimize(objective, callbacks=callbacks, n_trials=10, n_jobs=n_jobs)
+    assert values == [1 for _ in range(10)]
```
Comment (toshihikoyanase) [inline tests/test_study.py]: I think this assertion can be simplified like `assert values == [1] * 10`. We can see this style in [study.py](https://github.com/pfnet/optuna/blob/master/optuna/study.py#L456), so please think of rephrasing this line.

File: tests/test_study.py [hunk 3]
```diff
@@ -559,3 +559,58 @@ def test_storage_property():
...
+    # Two callbacks.
+    values = []
+    params = []
+    callbacks = [
+        lambda study, trial: values.append(trial.value),
+        lambda study, trial: params.append(trial.params)
+    ]
+    study.optimize(objective, callbacks=callbacks, n_trials=10, n_jobs=n_jobs)
+    assert values == [1 for _ in range(10)]
+    assert params == [{'x': 1} for _ in range(10)]
```
Comment (toshihikoyanase) [inline tests/test_study.py]: ditto.

File: tests/test_study.py (hunk 2 above)
Comment (sile) [reply tests/test_study.py]: Cool!
Fixed: https://github.com/pfnet/optuna/pull/480/commits/c3e0961a1181d5313fa34415adbc0feb99c30f2b

File: tests/test_study.py [hunk 4]
```diff
@@ -559,3 +559,58 @@ def test_storage_property():
...
+    assert params == [{'x': 1}] * 10
+
+    # If a trial is failed with an exception and the exception is caught by the study,
+    # callbacks are invoked.
+    states = []
+    callbacks = [lambda study, trial: states.append(trial.state)]
+    study.optimize(lambda t: 1/0, callbacks=callbacks, n_trials=10, n_jobs=n_jobs)
+    assert states == [optuna.structs.TrialState.FAIL] * 10
+
+    # NOTE: Because `Study.optimize` blocks forever if `n_jobs` is more than `1` and
```
Comment (g-votte) [inline tests/test_study.py]: I think we should leave a pointer to the original problem. I raised an issue; so, would you mind mentioning it in the comment?

https://github.com/pfnet/optuna/issues/538

Comment (sile) [reply tests/test_study.py]: Thanks. I agree with you.
Fixed: https://github.com/pfnet/optuna/pull/480/commits/aa930bd95d9f61c2ab20966fdb4a9925b0698fd0

**Pull request merged**